File: test/src/main/java/com/liangjz/test/test/design/chain/AbstractLogger.java
package com.liangjz.test.test.design.chain;

public abstract class AbstractLogger {
    public AbstractLogger mNextLogger;

    abstract void logger();

    public void startLogger(){
        this.logger();
        if(mNextLogger != null){
            mNextLogger.startLogger();
        }
    }

    public AbstractLogger nextLogger(AbstractLogger logger){
        this.mNextLogger = logger;
        return logger;
    }
}

File: test/src/main/java/com/liangjz/test/test/design/chain/kt/AbstractLogger.kt
package com.liangjz.test.test.design.chain.kt

abstract class AbstractLogger{
    var mNextLogger : AbstractLogger? = null

    abstract fun logger()

    fun startLogger(){
        this.logger()
        if(mNextLogger != null){
            mNextLogger!!.startLogger()
        }
    }

    fun nextLogger(next : AbstractLogger) : AbstractLogger{
        this.mNextLogger = next
        return next
    }

}